### PR TITLE
Project navbar

### PR DESCRIPTION
**Pull request by abought**
## Ticket
https://openscience.atlassian.net/browse/EOSF-98

# Purpose
Implement the project navbar, which wraps all project-related content.

For a contributor on a project:
screen shot 2016-07-14 at 11 09 26 pm

For a logged-in-user who is not a contributor on a (public) project:
screen shot 2016-07-18 at 5 24 39 pm

# Notes for Reviewers
- The `isProjectContributor` mechanism won't work until an API bug is fixed. This PR depends on: https://github.com/CenterForOpenScience/osf.io/pull/5981
- `showParentProjectLink` currently performs a separate fetch, pending fixes to request embeds (and handle embeds with error messages)
- View-only links have not been tested with this PR. 

## Future testing requirements
The future testing plan will require many combinations of access and resource type to ensure buttons display correctly. Partial manual testing has been performed, but was not exhaustive.
  - Project, component, registration
    - Active or withdrawn registrations
  - Public, private, view-only, anonymized view-only
  - Contributor or not
    - Read, write, or admin level permissions
  - Top level or child component
    - User can, or cannot, view parent project
- User can, or cannot, comment
- User has, or does not have, unread comments on this resource

# TODO
## Big-picture challenges
- We should support `embed=parent` for the node fetch request. This is how showing parent info is handled in the dashboard.
  - [x] Verify that navbar shows correct icon when parent project exists, but is private.
- Consider how we will handle API requests for view-only links on projects.
  - Verify that authentication + view only link returns the correct behavior
- [X] Be able to ask whether a given user ID is a contributor to a node. Doesn't use node.contributors, because it's limited to 10 items.
- [x] Add API property for whether a user is allowed to comment on a project. (OSF-6701)
- [x] Fix API bug with querying whether a user is allowed to contribute to a project. (OSF-6702)

## Specific items
- [x]  Provide a replacement for each variable checked in this template
- [ ] Get tooltip styles working for "parent component" arrow tooltip
- Remove knockout usages
- [x] Remove mako usages
- [ ] Fix style offset issue where navbar overlaps page content

## Routes Added/Updated
Add dummy navbar to node detail page, eg
`http://localhost:4200/nodes/<node_id>`

### CODE REVIEW

File: addon/components/eosf-project-nav/component.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return DS.PromiseObject.create({
+            promise: node.isContributor(userID)
+        });
+    }),
+
+    /**
+     * If this is a withdrawn registration, hide a block of buttons.
+     * @property minimalRegistrationView
+     */
+    minimalRegistrationView: Ember.computed.and('node.isRegistration', 'node.withdrawn'),
```
Comment (samchrisinger) [inline addon/components/eosf-project-nav/component.js]: Kind of nit-picky, but this computed's name seems to refer to the view state. Maybe `isWithdrawnRegistration` instead?

**Comment by samchrisinger** [comment]
@abought I'm going to hold on functional testing until the API blockers are on staging.

Are there any blockers for the remaining TODOs?

File: addon/components/eosf-project-nav/component.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if (node.get('isProject')) {
+            return node.get('currentUserPermissions').indexOf(permissions.WRITE) !== -1;
+        } else if (node.get('isRegistration')) {
+            return node.get('currentUserPermissions').indexOf(permissions.ADMIN) !== -1;
+        }
+        return false;  // No idea what this resource is, so don't show tab
+    }),
+
+    showCommentsButton: Ember.computed('node', function() {
+        // TODO: Implement. Depends on resolution of https://openscience.atlassian.net/browse/OSF-6701
```
Comment (abought) [inline addon/components/eosf-project-nav/component.js]: Will handle this in a followup PR- it makes more sense to turn on the comments pane button only once styling is actually in place.

File: addon/components/eosf-project-nav/component.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    minimalRegistrationView: Ember.computed.and('node.isRegistration', 'node.withdrawn'),
+
+    parentExists: Ember.computed('node', function() {
+        // Determine if a parent exists (field not empty), without trying to fetch it.
+        // This provides a way to use conditionals without errors that hang the page
+        let node = this.get('node');
+        return !!node.belongsTo('parent').link();
+    }),
+    showParentProjectLink: Ember.computed('node.parent', function() {
+        // TODO: Rewrite this to rely on embeds, which are more efficient
```
Comment (abought) [inline addon/components/eosf-project-nav/component.js]: Currently blocked by how embed functionality deals with API errors. (#EOSF-110)

In some cases, the user may have access to a child component but not be able to see the parent. In that case, `ember=parent` is an error response instead of a record.

File: addon/models/file.js [hunk 4]
```diff
@@ -42,5 +42,6 @@ export default OsfModel.extend(FileItemMixin, {
     // File attributes
     versions: DS.hasMany('file-version'),
     comments: DS.hasMany('comment'),
+    node: DS.belongsTo('node'),  // TODO: In the future apiv2 may also need to support this pointing at nodes OR registrations
```
Comment (abought) [inline addon/models/file.js]: Haven't delved into this item too much as yet; held off for now to limit scope creep.

File: addon/models/node.js [hunk 5]
```diff
@@ -74,6 +75,43 @@ export default OsfModel.extend(FileItemMixin, {
     }),
     logs: DS.hasMany('logs'),
 
+    // These are only computeds because maintaining separate flag values on different classes would be a headache TODO: Improve.
```
Comment (abought) [inline addon/models/node.js]: Very low priority at the moment.

**Comment by samchrisinger** [comment]
Also wondering about
> Get tooltip styles working for "parent component" arrow tooltip 

and 
> Fix style offset issue where navbar overlaps page content

**Comment by abought** [comment]
It looks like the blockers may be on staging, based on a very light initial check. Eg, this query works:
 https://staging-api.osf.io/v2/nodes/m9pv6/contributors/?filter[id]=7bauz

**Comment by abought** [comment]
> Also wondering about
> 
>     Get tooltip styles working for "parent component" arrow tooltip
> 
> and
> 
>     Fix style offset issue where navbar overlaps page content

We use bootstrap-style tooltips in the OSF that haven't been added to the addon yet. (prioritizing functionality over visual tweaks)
screen shot 2016-07-19 at 11 59 56 am

For the second point, see screenshots in project description; the navbar seems to overlap the heading "data" at the top of the page.

**Comment by samchrisinger** [comment]
@abought I just found this library: http://kushagragour.in/lab/hint/... seems preferable in some ways and perhaps more friendly to ember data-binding. I might take a stab at adding a component for this. Thoughts?

**Comment by abought** [comment]
Per @samchrisinger discussion: I've fixed the styling issue that was immediate blocker, and annotated the remainders. 

I'll create followup tickets for any other issues later today per triaging given above, but will go ahead and merge this to support efforts on the project contributors page.
